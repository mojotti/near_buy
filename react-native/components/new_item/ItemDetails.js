import React from 'react';
import { TextInput, View } from 'react-native';
import PropTypes from 'prop-types';
import { styles } from '../../src/static/styles/NewItemStyles';

export class ItemDetails extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      price: '',
      description: '',
    };

    this.handleTitleChange = this.handleTitleChange.bind(this);
    this.handleDescriptionChange = this.handleDescriptionChange.bind(this);
    this.handlePriceChange = this.handlePriceChange.bind(this);
  }

  handleTitleChange(text) {
    this.setState({ title: text });
    this.props.onTitleChange(text);
  }

  handleDescriptionChange(text) {
    this.setState({ description: text });
    this.props.onDescriptionChange(text);
  }

  handlePriceChange(text) {
    let newText = '';
    const numbers = '0123456789';

    for (let i = 0; i < text.length; i++) {
      if (numbers.indexOf(text[i]) > -1) {
        newText += text[i];
      }
    }

    this.setState({ price: newText });
    this.props.onPriceChange(this.state.price);
  }

  render() {
    return (
      <View>
        <View style={styles.itemDetailContainer}>
          <TextInput
            placeholder="Title"
            autoCapitalize="sentences"
            autoCorrect={false}
            autoFocus={this.state.mounted}
            maxLength={40}
            keyboardType="email-address"
            value={this.state.title}
            onChangeText={text => this.handleTitleChange(text)}
            style={[styles.itemDetails]}
            underlineColorAndroid="transparent"
          />
        </View>
        <View style={styles.itemDescriptionContainer}>
          <TextInput
            placeholder="Description"
            autoCapitalize="sentences"
            maxLength={400}
            autoCorrect={false}
            keyboardType="email-address"
            value={this.state.description}
            multiline={true}
            onChangeText={text => this.handleDescriptionChange(text)}
            style={[styles.itemDetails]}
            underlineColorAndroid="transparent"
          />
        </View>
        <View style={styles.itemDetailContainer}>
          <TextInput
            placeholder="Price"
            autoCorrect={false}
            autoFocus={false}
            keyboardType="numeric"
            maxLength={5}
            value={this.state.price}
            onChangeText={text => this.handlePriceChange(text)}
            style={[styles.itemDetails]}
            underlineColorAndroid="transparent"
          />
        </View>
      </View>
    );
  }
}

ItemDetails.propTypes = {
  onTitleChange: PropTypes.func.isRequired,
  onDescriptionChange: PropTypes.func.isRequired,
  onPriceChange: PropTypes.func.isRequired,
};
