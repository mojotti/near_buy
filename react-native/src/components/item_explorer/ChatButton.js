import React from 'react';
import { ActivityIndicator, Text, TouchableOpacity, View } from 'react-native';
import PropTypes from 'prop-types';
import FontAwesome from 'react-native-vector-icons/FontAwesome';
import { styles } from '../../static/styles/ItemDetailsStyles';


export default class ChatButton extends React.Component {
  _getIcon = () => {
    if (this.props.isCreatingChat) {
      return <ActivityIndicator size="large" color="#FFFFFF" />;
    }
    return <FontAwesome name="comments" size={50} color="#FFFFFF" />;
  };

  _getText = () => {
    if (this.props.isCreatingChat) {
      return 'Creating chat...';
    }
    return 'Chat with owner';
  };

  render() {
    return (
      <View>
        <TouchableOpacity
          style={styles.chatButton}
          onPress={this.props.onPress}
        >
          <View style={styles.chatButtonContainer}>
            {this._getIcon()}
            <Text style={styles.chatText}>{this._getText()}</Text>
          </View>
        </TouchableOpacity>
      </View>
    );
  }
}

ChatButton.propTypes = {
  onPress: PropTypes.func.isRequired,
  isCreatingChat: PropTypes.bool.isRequired,
};
