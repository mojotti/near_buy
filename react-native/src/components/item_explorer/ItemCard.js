import React from 'react';
import { Image, Text, TouchableOpacity, View } from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { styles } from '../../static/styles/ItemCardStyles';
import { baseStyles } from '../../static/styles/BaseStyles';
import { localhost } from '../../static/constants';
import { calculateDistanceInKm } from '../../utils/distance';

export class _ItemCard extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      distanceInKm: 'Getting distance...',
    };
  }

  componentDidMount() {
    this._getDistanceInKm(
      this.props.latitude,
      this.props.longitude,
      this.props.item.latitude,
      this.props.item.longitude
    );
  }

  componentDidUpdate(prevProps) {
    if (
      prevProps.item.latitude !== this.props.item.latitude ||
      prevProps.item.longitude !== this.props.item.longitude ||
      prevProps.latitude !== this.props.latitude ||
      prevProps.longitude !== this.props.longitude
    ) {
      this._getDistanceInKm(
        this.props.latitude,
        this.props.longitude,
        this.props.item.latitude,
        this.props.item.longitude
      );
    }
  }

  _getDistanceInKm = (lat1, lon1, lat2, lon2) => {
    const dist = calculateDistanceInKm(lat1, lon1, lat2, lon2);
    this.setState(() => ({ distanceInKm: `${dist} km` }));
  };

  _getImageSource = () => {
    return `http://${localhost}:5000/api/v1.0/${this.props.item.id}/image0.jpg`;
  };

  _navigateToItem = () => {
    const { navigate } = this.props.navigation;
    navigate('ItemDetails', {
      item: this.props.item,
      distance: this.state.distanceInKm,
    });
  };

  render() {
    return (
      <View style={styles.container}>
        <TouchableOpacity onPress={this._navigateToItem}>
          <Text style={baseStyles.headerText}>{this.props.item.title}</Text>
          <Image
            source={{ uri: this._getImageSource() }}
            style={styles.image}
          />
          <Text style={baseStyles.headerText}>
            {`Distance: ${this.state.distanceInKm}`}
          </Text>
        </TouchableOpacity>
      </View>
    );
  }
}

const mapStateToProps = state => {
  const { latitude, longitude } = state.locationReducer;
  return { latitude, longitude };
};

const ItemCard = connect(mapStateToProps, null)(_ItemCard);
export default ItemCard;

_ItemCard.propTypes = {
  item: PropTypes.shape({
    title: PropTypes.string,
    id: PropTypes.number,
    latitude: PropTypes.number.isRequired,
    longitude: PropTypes.number.isRequired,
  }).isRequired,
  latitude: PropTypes.number.isRequired,
  longitude: PropTypes.number.isRequired,
};
