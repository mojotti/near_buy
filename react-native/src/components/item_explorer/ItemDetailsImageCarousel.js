import React from 'react';
import { Dimensions } from 'react-native';
import PropTypes from 'prop-types';
import Carousel from 'react-native-looped-carousel';
import ImagePlaceholder from '../image_placeholder/ImagePlaceholder';
import { styles } from '../../static/styles/ItemDetailsImageCarouselStyles';

const { width } = Dimensions.get('window');


export default class ItemDetailsImageCarousel extends React.Component {
  render() {
    return (
      <Carousel
        style={styles.size}
        autoplay={false}
        pageInfo
        currentPage={0}
        pageInfoBackgroundColor={'#292929'}
        pageInfoTextStyle={{ color: '#FFFFFF' }}
      >
        {this.props.images.map((uri, i) => (
          <ImagePlaceholder
            url={uri}
            styles={styles.size}
            placeholderSize={width}
            key={i}
          />
        ))}
      </Carousel>
    );
  }
}

ItemDetailsImageCarousel.propTypes = {
  images: PropTypes.arrayOf(PropTypes.string).isRequired,
};
