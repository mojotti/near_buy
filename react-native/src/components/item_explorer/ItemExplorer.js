'use strict';

import React from 'react';
import { ListView, Text, TouchableOpacity, View } from 'react-native';
import { connect } from 'react-redux';
import Icon from 'react-native-vector-icons/MaterialIcons';

import { styles } from '../../static/styles/ItemsStyles';

export class ItemExplorer extends React.Component {
  static navigationOptions = ({ navigation }) => {
    const { navigate } = navigation;
    return {
      title: 'Item Explorer',
      headerLeft: (
        <View style={{ paddingHorizontal: 10 }}>
          <TouchableOpacity onPress={() => navigation.navigate('DrawerOpen')}>
            <Icon name="menu" size={30} color="blue" />
          </TouchableOpacity>
        </View>
      ),
      headerRight: (
        <Text onPress={() => navigate('NewItem')} style={styles.newItem}>
          New item
        </Text>
      ),
    };
  };

  render() {
    return (
      <View>
        <Text>Here will be items</Text>
      </View>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  return {};
};

export default connect(mapStateToProps, null)(ItemExplorer);
