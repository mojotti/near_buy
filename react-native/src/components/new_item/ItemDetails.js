import React from 'react';
import { TextInput, View } from 'react-native';
import PropTypes from 'prop-types';
import { styles } from '../../static/styles/NewItemStyles';

export class ItemDetails extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      price: '',
      description: '',
    };

    this.handleTitleChange = this.handleTitleChange.bind(this);
    this.handleDescriptionChange = this.handleDescriptionChange.bind(this);
    this.handlePriceChange = this.handlePriceChange.bind(this);
  }

  componentDidMount() {
    if (this.props.title && this.props.description && this.props.price) {
      this.setState(() => ({
        title: this.props.title,
        description: this.props.description,
        price: this.props.price.toString(),
      }));
    }
  }

  handleTitleChange(text) {
    this.setState({ title: text });
    this.props.onTitleChange(text);
  }

  handleDescriptionChange(text) {
    this.setState({ description: text });
    this.props.onDescriptionChange(text);
  }

  handlePriceChange(text) {
    let newText = '';
    const numbers = '0123456789';

    for (let i = 0; i < text.length; i++) {
      if (numbers.indexOf(text[i]) > -1) {
        newText += text[i];
      }
    }

    this.setState({ price: newText });
    this.props.onPriceChange(newText);
  }

  render() {
    return (
      <View>
        <View style={styles.itemDetailContainer}>
          <TextInput
            placeholder="Title"
            autoCapitalize="sentences"
            autoCorrect={false}
            maxLength={40}
            keyboardType="email-address"
            value={this.state.title}
            onChangeText={text => this.handleTitleChange(text)}
            style={[styles.itemDetails]}
            underlineColorAndroid="transparent"
          />
        </View>
        <View style={styles.itemDescriptionContainer}>
          <TextInput
            placeholder="Description"
            autoCapitalize="sentences"
            maxLength={400}
            autoCorrect={false}
            keyboardType="email-address"
            value={this.state.description}
            multiline={true}
            onChangeText={text => this.handleDescriptionChange(text)}
            style={[styles.itemDetails]}
            underlineColorAndroid="transparent"
          />
        </View>
        <View style={styles.itemDetailContainer}>
          <TextInput
            placeholder="Price"
            autoCorrect={false}
            autoFocus={false}
            keyboardType="numeric"
            maxLength={5}
            value={this.state.price}
            onChangeText={text => this.handlePriceChange(text)}
            style={[styles.itemDetails]}
            underlineColorAndroid="transparent"
          />
        </View>
      </View>
    );
  }
}

ItemDetails.defaultProps = {
  title: '',
  description: '',
  price: null,
};

ItemDetails.propTypes = {
  onTitleChange: PropTypes.func.isRequired,
  onDescriptionChange: PropTypes.func.isRequired,
  onPriceChange: PropTypes.func.isRequired,
  title: PropTypes.string,
  description: PropTypes.string,
  price: PropTypes.number,
};
