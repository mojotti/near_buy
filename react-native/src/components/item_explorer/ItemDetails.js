import React from 'react';
import { Text, ScrollView, View } from 'react-native';
import { connect } from 'react-redux';
import FontAwesome from 'react-native-vector-icons/FontAwesome';
import ItemDetailsImageCarousel from './ItemDetailsImageCarousel';
import { getNumOfPictures } from '../../networking/ApiCalls';
import { localhost } from '../../static/constants';
import { baseStyles } from '../../static/styles/BaseStyles';
import { styles } from '../../static/styles/ItemDetailsStyles';
import ChatButton from './ChatButton';
import ItemSeparator from '../user_items/ItemSeparator';
import { createChatAction } from '../../redux/actions/ChatActions';

window.navigator.userAgent = 'ReactNative';

const io = require('socket.io-client/dist/socket.io');
const connectionConfig = {
  jsonp: false,
  reconnection: true,
  reconnectionDelay: 100,
  reconnectionAttempts: 100000,
  transports: ['websocket'], // you need to explicitly tell it to use websockets
};

const path = `http://${localhost}:5000`;

export class _ItemDetails extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      imageUrls: ['foo.bar'], // provide some uri while loading actual img uris
    };

    this.socket = null;
    if (!this.socket || !this.socket.connected) {
      this.socket = io(path, connectionConfig);
      this.socket.on('connect', () => {
        console.log('connected!');
      });
      this.socket.emit('lol', 'Hello world!');
    }
  }

  static navigationOptions = ({ navigation }) =>
    ({ title: navigation.state.params.item.title });

  componentWillMount() {
    this.getImageUrls();
  }

  getImageUrls = () => {
    getNumOfPictures(this.props.item.id, this.props.token)
      .then((numOfPics) => {
        this.setImageUrls(numOfPics);
      });
  };

  setImageUrls = (numOfPics) => {
    const urls = [];
    for (let i = 0; i < numOfPics; i++) {
      const imagePath =
        `http://${localhost}:5000/api/v1.0/${this.props.item.id}/image${i}.jpg`;
      urls.push(imagePath);
    }
    this.setState(() => ({ imageUrls: urls }));
  };

  _openChat = () => {
    this.props.dispatch(createChatAction(
      this.props.item.seller_id,
      this.props.item.id,
      this.props.token,
    ));
  };

  render() {
    return (
      <ScrollView style={styles.container}>
        <ItemDetailsImageCarousel images={this.state.imageUrls} />
        <Text style={baseStyles.headerText}>Distance</Text>
        <View style={styles.locationIconContainer}>
          <FontAwesome name="map-marker" size={20} color={'#9f9f9f'} style={styles.locationIcon} />
          <Text style={styles.plainText}>{this.props.distance}</Text>
        </View>
        <ItemSeparator widthPercentage={0.86} />
        <Text style={baseStyles.headerText}>Description</Text>
        <Text style={styles.plainText}>{this.props.item.description}</Text>
        <ItemSeparator widthPercentage={0.86} />
        <ChatButton onPress={this._openChat} isCreatingChat={this.props.isCreatingChat} />
      </ScrollView>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  const { item } = ownProps.navigation.state.params;
  const { distance } = ownProps.navigation.state.params;
  const { token } = state.authorizationReducer;
  const { isLoading } = state.chatCreationReducer;

  return {
    item,
    token,
    distance,
    isCreatingChat: isLoading,
  };
};

const ItemDetails = connect(mapStateToProps, null)(_ItemDetails);
export default ItemDetails;
