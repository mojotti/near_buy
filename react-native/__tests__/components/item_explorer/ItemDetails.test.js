import React from 'react';
import { shallow } from 'enzyme';
import { _ItemDetails } from '../../../src/components/item_explorer/ItemDetails';

jest.mock('../../../src/networking/ApiCalls.js', () => {
  return {
    getNumOfPictures: () => Promise.resolve(2)
  };
});

const ITEM = {
  title: 'foo',
  id: 0,
  buyer_id: 0,
  seller_id: 0,
  description: 'bar',
};

const PROPS = {
  item: {
    title: 'foo',
    description: 'bar',
  },
  navigation: {
    state: {
      params: {
        item: {
          title: 'foo'
        }
      }
    }
  },
  token: 'fake token',
  distance: '2 km',
  isCreatingChat: false,
  error: null,
  id: 1
};
describe('<_ItemDetails />', () => {
  test('renders correctly', () => {
    fetch.mockResponseSuccess({ numOfImages: 2 });

    const wrapper = shallow(<_ItemDetails {...PROPS} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('chat is created', () => {
    const wrapper = shallow(
      <_ItemDetails
        isCreatingChat={false}
        error={null}
        token="foo"
        item={[]}
        distance="2 km"
        id={2}
      />
    );

    const result = wrapper.instance()._isChatCreated({ isCreatingChat: true });
    expect(result).toBeTruthy();
  });

  test('chat is not created', () => {
    let wrapper = shallow(
      <_ItemDetails
        isCreatingChat={false}
        error={null}
        token="foo"
        item={[]}
        distance="2 km"
        id={2}
      />
    );

    let result = wrapper.instance()._isChatCreated({ isCreatingChat: false });
    expect(result).toBeFalsy();

    wrapper = shallow(
      <_ItemDetails
        isCreatingChat
        error={undefined}
        token="foo"
        item={[]}
        distance="2 km"
        id={2}
      />
    );

    result = wrapper.instance()._isChatCreated({ isCreatingChat: true });
    expect(result).toBeFalsy();
  });

  test('opens chat on button press', () => {
    const dispatchSpy = jest.fn();
    const wrapper = shallow(
      <_ItemDetails
        isCreatingChat={false}
        error={null}
        token="foo"
        item={[]}
        distance="2 km"
        id={2}
        dispatch={dispatchSpy}
      />
    );

    wrapper.find('ChatButton').simulate('Press');
    expect(dispatchSpy.mock.calls).toMatchSnapshot();
  });

  test('navigates to chat on component did update when appropriate', () => {
    const naviSpy = jest.fn();
    const navigation = {
      navigate: naviSpy
    };

    const wrapper = shallow(
      <_ItemDetails
        navigation={navigation}
        isCreatingChat={false}
        error={null}
        token="foo"
        item={ITEM}
        distance="2 km"
        id={2}
        dispatch={() => {}}
      />
    );

    wrapper.instance().componentDidUpdate({ isCreatingChat: true });
    const expected = [['Chat', { item: { buyer_id: 2, id: 0, seller_id: 0, title: 'foo' } }]];

    expect(naviSpy.mock.calls).toEqual(expected)
  });
});
