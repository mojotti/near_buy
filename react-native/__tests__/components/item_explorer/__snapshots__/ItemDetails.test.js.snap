// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<_ItemDetails /> opens chat on button press 1`] = `
Array [
  Array [
    [Function],
  ],
]
`;

exports[`<_ItemDetails /> renders correctly 1`] = `
<ScrollView
  style={
    Object {
      "backgroundColor": "#f5f5f5",
      "flex": 1,
    }
  }
>
  <ItemDetailsImageCarousel
    images={
      Array [
        "foo.bar",
      ]
    }
  />
  <Component
    style={
      Object {
        "alignSelf": "stretch",
        "color": "#848484",
        "fontFamily": null,
        "fontSize": 25,
        "marginBottom": 15,
        "marginLeft": 56.25,
        "marginTop": 20,
        "textAlign": "left",
      }
    }
  >
    Distance
  </Component>
  <Component
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "flex-start",
      }
    }
  >
    <Icon
      allowFontScaling={false}
      color="#9f9f9f"
      name="map-marker"
      size={20}
      style={
        Object {
          "marginLeft": 75,
          "marginRight": -67.5,
        }
      }
    />
    <Component
      style={
        Object {
          "alignSelf": "stretch",
          "color": "#2b2b2b",
          "fontFamily": null,
          "fontSize": 16,
          "marginBottom": 30,
          "marginLeft": 75,
          "marginRight": 75,
          "textAlign": "left",
        }
      }
    >
      2 km
    </Component>
  </Component>
  <ItemSeparator
    widthPercentage={0.86}
  />
  <Component
    style={
      Object {
        "alignSelf": "stretch",
        "color": "#848484",
        "fontFamily": null,
        "fontSize": 25,
        "marginBottom": 15,
        "marginLeft": 56.25,
        "marginTop": 20,
        "textAlign": "left",
      }
    }
  >
    Description
  </Component>
  <Component
    style={
      Object {
        "alignSelf": "stretch",
        "color": "#2b2b2b",
        "fontFamily": null,
        "fontSize": 16,
        "marginBottom": 30,
        "marginLeft": 75,
        "marginRight": 75,
        "textAlign": "left",
      }
    }
  >
    bar
  </Component>
  <ItemSeparator
    widthPercentage={0.86}
  />
  <ChatButton
    isCreatingChat={false}
    onPress={[Function]}
  />
</ScrollView>
`;
